File: DKPlugins/DKConsole/DKConsoleDUK.cpp
/*
* This source file is part of digitalknob, the cross-platform C/C++/Javascript/Html/Css Solution
*
* For the latest information, see https://github.com/aquawicket/DigitalKnob
*
* Copyright(c) 2010 - 2023 Digitalknob Team, and contributors
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files(the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and /or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions :
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

#ifdef HAVE_DKDuktape 
#include "DK/DKApp.h"
#include "DKConsole/DKConsoleDUK.h"


bool DKConsoleDUK::Init(){
	DKDEBUGFUNC();
	
	////// Instance methods //////
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_assert",		DKConsoleDUK::_assert); 		// [console.assert()] https://developer.mozilla.org/en-US/docs/Web/API/console/assert
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_clear", 		DKConsoleDUK::clear);			// [console.clear()] https://developer.mozilla.org/en-US/docs/Web/API/console/clear
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_count", 		DKConsoleDUK::count);			// [console.count()] https://developer.mozilla.org/en-US/docs/Web/API/console/count
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_countReset", 	DKConsoleDUK::countReset);		// [console.countReset()] https://developer.mozilla.org/en-US/docs/Web/API/console/countReset
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_debug", 		DKConsoleDUK::debug);			// [console.debug()] https://developer.mozilla.org/en-US/docs/Web/API/console/debug
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_dir", 			DKConsoleDUK::dir);				// [console.dir()] https://developer.mozilla.org/en-US/docs/Web/API/console/dir
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_error", 		DKConsoleDUK::error);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_exception", 	DKConsoleDUK::exception);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_group", 		DKConsoleDUK::group);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_groupCollapsed",DKConsoleDUK::groupCollapsed);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_groupEnd", 		DKConsoleDUK::groupEnd);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_info", 			DKConsoleDUK::info);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_log", 			DKConsoleDUK::log);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_trace", 		DKConsoleDUK::trace);
	DKDuktape::AttachFunction("CPP_DKConsoleDUK_warn", 			DKConsoleDUK::warn);
	
	DKClass::DKCreate("DKConsole/DKConsoleDUK.js");
	return true;
}


////// Instance methods //////
// [console.assert()] https://developer.mozilla.org/en-US/docs/Web/API/console/assert
int DKConsoleDUK::_assert(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKERROR(string);
	return true;
}
// [console.clear()] https://developer.mozilla.org/en-US/docs/Web/API/console/clear
int DKConsoleDUK::clear(duk_context* ctx){
	int rtnvalue = DKUtil::System("cls", rtnvalue);
	duk_push_int(ctx, rtnvalue);
	return true;
}
// [console.count()] https://developer.mozilla.org/en-US/docs/Web/API/console/count
int DKConsoleDUK::count(duk_context* ctx){
	return DKTODO();
}
// [console.countReset()] https://developer.mozilla.org/en-US/docs/Web/API/console/countReset
int DKConsoleDUK::countReset(duk_context* ctx){
	return DKTODO();
}
// [console.debug()] https://developer.mozilla.org/en-US/docs/Web/API/console/debug
int DKConsoleDUK::debug(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKDEBUG(string);
	return true;
}
// [console.dir()] https://developer.mozilla.org/en-US/docs/Web/API/console/dir
int DKConsoleDUK::dir(duk_context* ctx){
	return DKTODO();
}

int DKConsoleDUK::error(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKERROR(string);
	return true;
}

int DKConsoleDUK::exception(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKERROR(string);
	return true;
}

int DKConsoleDUK::group(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKINFO(string);
	return true;
}

int DKConsoleDUK::groupCollapsed(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKINFO(string);
	return true;
}

int DKConsoleDUK::groupEnd(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKINFO(string);
	return true;
}

int DKConsoleDUK::info(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKINFO(string);
	return true;
}

int DKConsoleDUK::log(duk_context* ctx){
	DKString string;
	if (duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if (duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if (duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKINFO(string);
	return true;
}

int DKConsoleDUK::trace(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKWARN(string);
	return true;
}

int DKConsoleDUK::warn(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0))
		string = duk_require_string(ctx, 0);
	if(duk_is_boolean(ctx, 0))
		string = toString(duk_require_boolean(ctx, 0));
	if(duk_is_number(ctx, 0))
		string = toString(duk_require_int(ctx, 0));
	DKWARN(string);
	return true;
}


#endif //HAVE_DKDuktape

File: DKPlugins/DKConsole/DKConsoleDUK.js
// [IDL] https://console.spec.whatwg.org/#console-namespace
// [MDN] https://developer.mozilla.org/en-US/docs/Web/API/Console


//[Exposed=*]
//namespace console { // but see namespace object requirements below
var Console = function Console(address) {
	//CPP_DKConsoleDUK_log("Console("+address+")");
	
	if(address)
		this.address = address;
	if(!this.address)
		this.address = CPP_DKConsoleDUK_constructor();
	
	
	// Logging
	//undefined assert(optional boolean condition = false, any... data);
    Console.prototype.assert = function assert(condition, data) {
        CPP_DKConsoleDUK_assert(this.address, condition, data);
    }
	
	//undefined clear();	
    Console.prototype.clear = function clear() {
        CPP_DKConsoleDUK_clear(this.address);
    }
	
	//undefined debug(any... data);
	Console.prototype.debug = function debug(data) {
        CPP_DKConsoleDUK_debug(this.address, data);
    }
	
	//undefined error(any... data);
	Console.prototype.error = function error(data) {
		/*
		const err = new Error(data);
		if (!err.stack) {
			try {
				throw err; // old browsers need the Error thrown to fill the stack
			} 
			catch (err) {
				if (!err.stack)
					return error("err.stack invalid"); // browser too old
			}
		}
		*/
		CPP_DKConsoleDUK_error(this.address, data);
    }
	
	//undefined info(any... data);
	Console.prototype.info = function info(data) {
        CPP_DKConsoleDUK_info(this.address, data);
    }
	
	//undefined log(any... data);
	Console.prototype.log = function log(data) {
        CPP_DKConsoleDUK_log(this.address, data);
    }
	
	//undefined table(optional any tabularData, optional sequence<DOMString> properties);
	Console.prototype.table = function table(tabularData, properties) {
        CPP_DKConsoleDUK_table(this.address, tabularData, properties);
    }
	
	//undefined trace(any... data);
	Console.prototype.trace = function trace(data) {
		CPP_DKConsoleDUK_trace(this.address, data);
    }
	
	//undefined warn(any... data);
	Console.prototype.warn = function warn(data) {
        CPP_DKConsoleDUK_warn(this.address, data);
    }
	
	//undefined dir(optional any item, optional object? options);
	Console.prototype.dir = function dir(item, options) {
        CPP_DKConsoleDUK_dir(this.address, item, options);
    }
	
	//undefined dirxml(any... data);
	Console.prototype.dirxml = function dirxml(data) {
        CPP_DKConsoleDUK_dirxml(this.address, data);
    }
	
	// Counting
	//undefined count(optional DOMString label = "default");
	Console.prototype.count = function count(label) {
        CPP_DKConsoleDUK_count(this.address, label);
    }
	
	//undefined countReset(optional DOMString label = "default");
	Console.prototype.countReset = function countReset(label) {
        CPP_DKConsoleDUK_dirxml(this.address, label);
    }
   
	// Grouping
	//undefined group(any... data);
	Console.prototype.group = function group(data) {
        CPP_DKConsoleDUK_group(this.address, data);
    }
	
	//undefined groupCollapsed(any... data);
	Console.prototype.groupCollapsed = function groupCollapsed(data) {
        CPP_DKConsoleDUK_groupCollapsed(this.address, data);
    }
	
	//undefined groupEnd();
	Console.prototype.groupEnd = function groupEnd() {
        CPP_DKConsoleDUK_groupEnd(this.address);
    }
	
	// Timing
	//undefined time(optional DOMString label = "default");
	Console.prototype.time = function time(label) {
        CPP_DKConsoleDUK_time(this.address, label);
    }
	
	//undefined timeLog(optional DOMString label = "default", any... data);
	Console.prototype.timeLog = function timeLog(label, data) {
        CPP_DKConsoleDUK_timeLog(this.address, label, data);
    }
	
	//undefined timeEnd(optional DOMString label = "default");
	Console.prototype.timeEnd = function timeEnd(label) {
        CPP_DKConsoleDUK_timeEnd(this.address, label);
    }
	   
	
	////// toString //////
	//NOTE: Chrome returns [object Object]
	if(this.toString() === "[object Object]")
		this.toString = function(){ return "[object Console]" }
	
	return this;
}

var console = new Console();
